### ISSUE TITLE
`node:test` is not considered a builtin module by `import/order`

### ISSUE THREAD

**Issue by remcohaszing** [import-js/eslint-plugin-import#2603]
When using the `import/order` rule, I expect the module `node:test` to be treated as a builtin module, but it’s treated as an external module instead.

I tried to see if this is an issue in `is-core-module`, but it returns the correct result.

```js
> require('is-core-module')('node:test')
true
```

**Comment by ljharb** [comment]
Absolutely the node version determines whether it's a core module.

The proper solution is to configure vscode to use the node version you want, instead of its builtin one.

In older versions of node, `node:`-prefixed modules aren't always core modules - they could come from the filesystem - and also, webpack or typescript could still provide them.

**Comment by remcohaszing** [comment]
I agree that the Node.js version determines whether it’s a core module, but I don’t think the Node.js version used to run ESLint should be used to determine the ESLint output.

Indeed VSCode can be configured to use a different node executable. I have configured this on my end now, thanks for the tip. Still, I think this is a workaround, not a proper solution.

I wasn’t aware `node:`-prefixed modules can be non-core modules in older version of Node.js. Also I guess non-prefixed modules can still be introduced nowadays.

Module bundlers can provide any module, so by saying Webpack can provide a module, no module can be considered builtin.

[`eslint-plugin-n`](https://github.com/eslint-community/eslint-plugin-n) has similar issues. Their rules accept the [`version`](https://github.com/eslint-community/eslint-plugin-n/blob/master/docs/rules/no-unsupported-features/es-builtins.md#version) option. I think it makes sense to support that here too.

**Comment by ljharb** [comment]
Sure, and that's defined by a custom resolver - the webpack resolver can indicate what's a builtin however it likes. The node resolver, however, does it based on the current node version.

I agree it would be possible to add a node resolver config option to specify the node version; I'm not yet convinced it's a good idea, since the only use case I'm aware of is vscode, since vscode chooses to ship its own bundled node.

**Comment by remcohaszing** [comment]
Let’s say a package uses a module introduced in a newer version of Node.js. To support older Node.js versions, a polyfill is added as a dependency. Contributors can run the test suite using either Node.js version and it works. However, this ESLint plugin now produces different results depending on the Node.js version used by the contributor.

**Comment by ljharb** [comment]
ok, hmm. I was thinking about the use cases here.

What I can imagine is something like a "use engines" flag in the node resolver config. When enabled:
 - the `order` rule would pin the core module resolution to the latest (or earliest, perhaps configurably) supported node version in the package's engines.node range
 - the `no-unresolved` rule would check *every* node version in the `engines.node` range to make sure a module resolves as *something* in all of them.
 - TODO: figure out what makes sense for any other rules that use the resolver

That way, you wouldn't have to configure _which_ node version to check, it'd just "do the right thing" for each relevant rule.

Would that meet your use cases, even without the ability to explicitly specify a node version specifically for eslint?

**Comment by remcohaszing** [comment]
I like where this is going. It makes sense to add this option to a resolver. (This is still configured in `.eslintrc`, right?

> * the `order` rule would pin the core module resolution to the latest (or earliest, perhaps configurably) supported node version in the package's engines.node range

I think it would be better to pick an exact version, not a range. Otherwise I don’t see how to determine which version to target. A user might not specify `engines.node`. Also newer Node.js versions can be released at any time.

> * the `no-unresolved` rule would check _every_ node version in the `engines.node` range to make sure a module resolves as _something_ in all of them.

Just like the point above, this means behaviour would change when a new Node.js version is released.

> * TODO: figure out what makes sense for any other rules that use the resolver

I think this affects [`import/no-nodejs-modules`](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-nodejs-modules.md).

**Comment by remcohaszing** [comment]
I’m not sure how you’d know a newer version of Node.js is released. Anyway, apart from that I like the proposed solution.

**Comment by brettz9** [comment]
> The proper solution is to configure vscode to use the node version you want, instead of its builtin one.

I'm getting a different linting result on the command line and from VSC here, despite having used nvm to set v20.0.0 as the default (in my case for `import/no-unresolved`). I've tried different solutions to indicate to VSC what the Node version is (on a Mac), but none have worked. A ny knowledge of how to fix? Thanks!

**Comment by ljharb** [comment]
@brettz9 searching through vscode's settings implies that the `"eslint.nodePath": ""` setting might be helpful.

**Comment by ljharb** [comment]
You may want to file an issue with vscode, then - i'm not sure how to change it.
